optimize/cmaes.go

```diff
@@ -226,90 +230,165 @@ func (cma *CmaEsChol) InitGlobal(dim, tasks int) int {
+}
 
-	// Get the next task and send it to be run if there is a next task to be run.
-	// If all of the tasks have been run, perform an update step. Note that the
-	// use of this mutex means that only one task can proceed, all of the
-	// other tasks should get stuck and then get a new location.
-	cma.mux.Lock()
-	if len(cma.evals) != 0 {
-		// There are still tasks to evaluate. Grab one and remove it from the list.
-		newIdx := cma.evals[len(cma.evals)-1]
-		cma.evals = cma.evals[:len(cma.evals)-1]
-		cma.wg.Add(1)
-		cma.mux.Unlock()
-
-		// Sample x and send it to be evaluated.
-		distmv.NormalRand(cma.xs.RawRowView(newIdx), cma.mean, &cma.chol, cma.Src)
-		copy(loc.X, cma.xs.RawRowView(newIdx))
-		cma.taskIdxs[task] = newIdx
-		return FuncEvaluation, nil
-	}
-	// There are no more tasks to evaluate. This means the iteration is over.
-	// Find the best current f, update the parameters, and re-establish
-	// the evaluations to run.
+// sendTask generates a sample and sends the task. It does not update the cma index.
+func (cma *CmaEsChol) sendTask(idx int, task GlobalTask) {
+	task.ID = idx
+	task.Op = FuncEvaluation
+	distmv.NormalRand(cma.xs.RawRowView(idx), cma.mean, &cma.chol, cma.Src)
+	copy(task.X, cma.xs.RawRowView(idx))
+	cma.operation <- task
+}
 
-	// Wait for all of the outstanding tasks to finish, so the full set of functions
-	// has been evaluated.
-	cma.wg.Wait()
+// bestIdx returns the best index in the functions. Returns -1 if all values
+// are NaN.
+func (cma *CmaEsChol) bestIdx() int {
+	best := -1
+	bestVal := math.Inf(1)
+	for i, v := range cma.fs {
+		if math.IsNaN(v) {
+			continue
+		}
+		if v < bestVal {
+			best = i
+			bestVal = v
+		}
+	}
+	return best
+}
 
-	// Find the best f out of all the tasks.
-	best := floats.MinIdx(cma.fs)
+// findBestAndUpdateTask finds the best task in the current list, updates the
+// new best overall, and then stores the best location into task.
+func (cma *CmaEsChol) findBestAndUpdateTask(task GlobalTask) GlobalTask {
+	// Find and update the best location.
+	// Don't use floats because there may be NaN values.
+	best := cma.bestIdx()
 	bestF := cma.fs[best]
 	bestX := cma.xs.RawRowView(best)
 	if cma.ForgetBest {
-		loc.F = bestF
-		copy(loc.X, bestX)
+		task.F = bestF
+		copy(task.X, bestX)
 	} else {
 		if bestF < cma.bestF {
 			cma.bestF = bestF
 			copy(cma.bestX, bestX)
 		}
-		loc.F = cma.bestF
-		copy(loc.X, cma.bestX)
-	}
+		task.F = cma.bestF
+		copy(task.X, cma.bestX)
 
-	cma.taskIdxs[task] = -1
-
-	// Update the parameters of the distribution
-	err := cma.update()
+	}
+	return task
+}
 
-	// Reset the tasks
-	cma.evals = cma.evals[:cma.pop]
+func (cma *CmaEsChol) RunGlobal(operation chan<- GlobalTask, result <-chan GlobalTask, tasks []GlobalTask) {
+	cma.operation = operation
+	// Send the initial tasks. We know there are at most as many tasks as elements
+	// of the population.
+	cma.sendInitTasks(tasks)
+	cma.taskIdx = len(tasks)
+
+Outer:
+	for {
+		task := <-result
```

Comment:
Seeing the implementation I think I would globally rename the `result` channel to `results`, and so `operation` to `operations`. Then here `result := <-results` and below `task := findBestAndUpdateTask(result)`.

Reply:
Done.